lib/ask/runtime/survey.ex

```diff
@@ -43,138 +43,103 @@ defmodule Ask.Runtime.Survey do
+                      |> Flow.resulting_disposition(reply_disposition)
+                      |> Flow.resulting_disposition("completed")
+
+    updated_respondent = respondent_updates(:end, respondent, new_disposition, offline)
+                         |> disposition_changed(session, old_disposition, offline)
+
+    # If new_disposition == reply_disposition, change of disposition has already been logged during Session.sync_step
+    if offline && session && new_disposition != old_disposition && new_disposition != reply_disposition do
+      mode = session.current_mode |> SessionMode.mode
+      Session.log_disposition_changed(updated_respondent, session.current_mode.channel, mode, old_disposition, new_disposition, offline)
+    end
 
     case Reply.steps(reply) do
       [] ->
-        :end
+        {:end, updated_respondent}
       _ ->
-        {:end, {:reply, reply}}
+        {:end, {:reply, reply}, updated_respondent}
     end
   end
 
-  def handle_session_step({:rejected, reply, respondent}, _) do
-    update_respondent(respondent, :rejected)
-    {:end, {:reply, reply}}
-  end
-
-  def handle_session_step({:rejected, %{respondent: respondent} = session, reply, timeout}, _) do
-    update_respondent(respondent, {:rejected, session, timeout})
-    {:reply, reply}
-  end
-
-  def handle_session_step({:rejected, respondent}, _) do
-    update_respondent(respondent, :rejected)
-    :end
-  end
-
-  def handle_session_step({:stopped, reply, respondent}, _) do
-    update_respondent(respondent, :stopped, Reply.disposition(reply), nil)
-    :end
+  def handle_session_step({:rejected, reply, respondent}, _, offline) do
+    updated_respondent =respondent_updates(:rejected, respondent, nil, nil, offline)
+    {:end, {:reply, reply}, updated_respondent}
   end
 
-  def handle_session_step({:failed, respondent}, _) do
-    update_respondent(respondent, :failed)
-    :end
+  def handle_session_step({:rejected, %{respondent: respondent} = session, reply, timeout}, _, offline) do
+    timeout_at = Respondent.next_actual_timeout(respondent, timeout, SystemTime.time.now, offline)
+    updated_respondent = respondent_updates(:rejected, respondent, session, timeout_at, offline)
+    {:reply, reply, updated_respondent}
   end
 
-  def update_respondent(%Respondent{} = respondent, :end) do
-    update_respondent(respondent, :end, nil, nil)
+  def handle_session_step({:rejected, respondent}, _, offline) do
+    updated_respondent = respondent_updates(:rejected, respondent, nil, nil, offline)
+    {:end, updated_respondent}
   end
 
-  def update_respondent(%Respondent{} = respondent, :rejected) do
-    respondent
-    |> Respondent.changeset(%{state: "rejected", session: nil, timeout_at: nil})
-    |> Repo.update!
+  def handle_session_step({:stopped, reply, respondent}, _, offline) do
+    updated_respondent = respondent_updates(:stopped, respondent, Reply.disposition(reply), offline) # TODO: update_respondent_and_set_disposition
```

Comment:
Is the `update_respondent_and_set_disposition` still pending? We should not forget about that.

Reply:
no, thanks!